### ISSUE TITLE
Transformations architecture

### ISSUE THREAD

**Issue by jni** [Napari/napari#3783]
## 🚀 Feature

This is a brain dump of ideas that have been floating around in my head for a while and that I have mentioned in passing in various contexts but never written down as a dedicated issue for discussion. At some point I will probably need to formally write it up in an @andy-sweet -style design doc — should we call them Sweet Docs from now on? :joy: 

## Motivation / current situation

Currently we have (at least) four explicit coordinate spaces, four implicit spaces, and a mishmash of methods on different components (layers, viewer, cursors, VisPy layers...) to transform between them.

The explicit spaces:
- data
- physical
- world
- grid

The implicit spaces:
- tile (one before data)
- worldslice
- "vispy world" — like worldslice but xyz instead of zyx, see e.g. #3767
- camera/viewbox

*Specifying* the data-to-physical transform is done by passing the scale, translate, rotation, and shear arguments to the layer. The physical to world transform is specified by passing the affine argument to the layer.

Already I find this problematic: there's nothing to say that a user shouldn't be able to specify a data-to-physical transformation as an affine matrix (and some LLS users might prefer this), or that the transform shouldn't be more complicated (as in the case of lens aberration correction, for example). The choice of affine as the physical-to-world transformation is similarly motivated by convenience (it is the most complex transformation that vispy supports out of the box), not by any fundamental reason to limit that space.

As to how one *accesses* the transforms, this is currently a mishmash of `layer.data_to_world`, `layer._world_to_data` (!), and `layer.transforms[1:3]` (for example). The indices into the transform chain one has to basically know from memory.

Slicing is even hairier: there we exit the transforms machinery completely and rely on various bespoke invocations of slicing logic involving `layer._dims_displayed`, `layer._dims_not_displayed`, and `layer._dims_displayed_order`. (Or similar, I might have the names wrong here.)

## Pitch / vision

I'm a bit murky on the details, or this would be a PR. :joy: But here are some ideas to aspire to:

- everything is part of the transform chain, even slicing and viewing.
- elements add to the transform chain as the transforms they contribute become available. So, for example, layers have transforms from data to world, then sliced layers add world to worldslice, and then the canvas adds canvas to worldslice, and finally VisPy adds canvas to vispy canvas. Whenever an element is added, the chain is copied, so that the full transform chain for each layer is available on the canvas.
- all *spaces* are named (data, physical, world, etc) (maybe a StringEnum here?), and transform chain objects can return both the *extent* of the chain (e.g. "I go from 'data' to 'worldslice' and everywhere in between"), and arbitrary slices of it with `layer.transforms['data':'world']` or similar syntax.
- all transforms, other than the projection from 3D to camera, must be invertible, so that the cursor can always be traced back to the data.
- users and plugins can specify whatever transform they like to put on the chain, as long as it satisfies invertibility *and* provides VisPy glsl map and imap code, as done in [visvis](https://github.com/codypiersall/vismap/blob/9a7e44bd19ddf50550ef7d8d89a260fd3bd50931/vismap/transforms.py).
- *Slicing* of an n-dimensional transform can initially be done ignoring non-orthogonal dims as is done now, but could in fact be more sophisticated. For example, if you have a z/y shear and look at xy, the translation on the sliced layer could be updated on each slice. On the VisPy side, it's a bit hairier but I think it can be done by generating the glsl dynamically during slicing. (We'll need this if we want to allow out-of-memory non-orthogonal slicing.)

## Alternatives

Continue plodding along haphazardly. :joy: 

## Additional context

- napari for geodata: #973 
- the discussion in the ome-ngff issue on transformations [also notes](https://github.com/ome/ngff/pull/57#issuecomment-939985686) the use of separate "physical" and "world" coordinate spaces and its utility for registration, for example, so I'm happy we rediscovered the same principle. =)
- BigDataViewer and BigVolumeViewer are good models for how to think about transformations. They have solved this problem pretty much as described above (per my understanding, anyway), but they are CPU-based so they have a slightly easier job. =)

**Comment by jni** [comment]
I've made a "vision" tag for "big picture" issues, because I've started to realise that a lot of that knowledge is internal to a few of us, or scattered across many smaller issues, and we need to do a better job of communicating the overall direction. Having just a few of these vision issues, easily collated, might really help everyone be on the same page about where we want the software to go!

**Comment by nclack** [comment]
> BigDataViewer and BigVolumeViewer are good models

strong agree

> "vision" tag for "big picture"

This is maybe a separate discussion, but I love the idea of something like [NEPs][].  I think that kind of thing is fantastic for guiding discussions into some concrete piece of useful documentation.  @andy-sweet has done some great work along this direction imo.

[neps]: https://numpy.org/neps/

> everything is part of the transform chain, even slicing and viewin

I love this. _especially_ for slicing.  We basically aways want to be able to map a query point, r, back into the originating data space.  The query point might be a 2d point in the view but we can always map that back into n dimensional data. For example (non affine for simplicity):

```
r_data =    T * r_query
 [Nx1] = [Nx2] [2x1]
```

> must be invertible

I'm not so sure about this. I think allowing non-invertable transforms (esp. projections) will be important for tracking dims.  I'm interested in hearing more!

>  non-orthogonal slicing.

I really want to look more into BigDataViewer to see how this is handled. I think it's pretty important. 

Not sure of the right way to handle this, but I like to think that the "slicing" function could accept non-orthogonal box queries, and then return orthogonal data - something different than the query. The returned data would be accompanied by a description of its extent and, of course, a transform.  I think this idea needs a bit more flushing out. It's something I've discussed w @perlman.

**Comment by tlambert03** [comment]
> This is maybe a separate discussion, but I love the idea of something like NEPs.

I guess we should call them ... NEPs ??

> @andy-sweet has done some great work along this direction imo.

💯.  role model :)

**Comment by jni** [comment]
> I guess we should call them ... NEPs ??

In informal discussion we suggested NAPs, "napari acceleration/augmentation/awesome proposal". 😜 I'm not that fussed but it should definitely *not* be NEP.

Anyway, part of the role of NEPs is to summarise discussion in issues like this and come to a decision, so I see them as additional to the vision tag, not a replacement for it.

> > must be invertible
>
> I'm not so sure about this.

Well, you *have* to have a way to go from canvas to data, or you can't actually draw, right? Or are you talking about MIP/other 3D rendering? In which case I'm thinking (not sure on this) that the GL depth buffer is actually part of the transform in some way.

> I really want to look more into BigDataViewer to see how this is handled

I *thought* BDV was GPL, but it seems either I was incorrect all along or they've split out the core to be BSD, which is *great news*(/not news 😂): https://github.com/bigdataviewer

Anyway, Stephan Saalfeld gave me a quick overview of the method when he visited 2y ago (The Good Ol' Days), reproducing here from memory so there may be gaps/inaccuracies:
- it's CPU only: basically 3000x2000 covers most monitors and is a small number of pixels to draw.
- they have a transform chain similar to that described above. They then iterate over the pixels on screen, follow the transform chain to the data, and load that data point.
- all of the operations are JITted so this is "fast" per-pixel
- The data is backed by ImgLib2, which handles chunking and caching transparently, so that loading adjacent data points during a screen refresh is fast even if the backing array (storage/compute) is slow

We have a more complicated task because (a) we don't have imglib2 (dask fills some but definitely not all of that role), and (b) we have a two step rendering pipeline, one loading the data onto the GPU and the rest on the GPU.

**Comment by nclack** [comment]
> you have to have a way to go from canvas to data

agreed. It's the other direction (data to canvas) I'm not sure we need.

**Comment by jojoelfe** [comment]
One thing I really would like to be able to do is instead of using an nD affine matrix for an nD dataset is to provide a list of (n-k)D affine matrices along k dimensions. The main usecase I have in mind is video-stabilization, where one would assign a 2D affine to every frame of a movie. Another usecase are montage-datasets where one could assign a 2D affine to every tile. Together with a way to "slab" views along a dimension ( #3600 ) this would provide a way to quickly assemble montages inside napari.

**Comment by jni** [comment]
> agreed. It's the other direction (data to canvas) I'm not sure we need.

Ah, sure, but (a) it's usually more natural to provide that direction, and (b) usually projections (where this would be an issue) happen on the way to the canvas, not the other way around?

Anyway, I should say I'm also not sure on this, but it seems like a reasonable constraint that will make our lives easier. 😂 

> One thing I really would like to be able to do is instead of using an nD affine matrix for an nD dataset is to provide a list of (n-k)D affine matrices along k dimensions

Yes, this would be feasible with the above framework: you'd "just" provide a "PiecewiseAffine" transform with the corresponding data. We also want this for e.g. aligned stacks of EM data. Dealing with the transforms on the GPU is another issue 😅, but I don't think an insurmountable one. (I *think* it would involve storing the transform parameters as a texture.)

**Comment by perlman** [comment]
> everything is part of the transform chain, even slicing and viewing.

Yes, and this would extend down to individual tiles(/chunks) in a tiled(/chunked) datasource... Tiling is done by adding a translation to the chain for each tile.

> Not sure of the right way to handle this, but I like to think that the "slicing" function could accept non-orthogonal box queries, and then return orthogonal data - something different than the query.

I am currently thinking of "slicing" as a rectangular matrix, where the output dimensions are fewer than input.

(Edit: I believe it was @andy-sweet that mention rectangular matrices in this weeks architecture working group.)

**Comment by jni** [comment]
> I am currently thinking of "slicing" as a rectangular matrix, where the output dimensions are fewer than input.

can you elaborate on how this works with invertibility? I don't have a good mental model of how slicing works with the matrices and how you can infer the original high-D coordinate from the low-D coordinate + the matrix...?

**Comment by andy-sweet** [comment]
> (Edit: I believe it was @andy-sweet that mention rectangular matrices in this weeks architecture working group.)

Obviously it won't work with invertibility without other information (as is true in general when going from 2D canvas to nD data/world coordinates). This initially came up around a potential fix for #3410 where we take a square submatrix when slicing a transform, which loses information about the source/domain of that transform, whereas a rectangular submatrix would not. Given the discussion here is much broader, I think we can mostly ignore it.

**Comment by andy-sweet** [comment]
> I really want to look more into BigDataViewer to see how this is handled. I think it's pretty important.

Agreed. As we seem to be using these as exemplars, the sooner we can reference code or design docs, the better.

**Comment by andy-sweet** [comment]
> One thing I really would like to be able to do is instead of using an nD affine matrix for an nD dataset is to provide a list of (n-k)D affine matrices along k dimensions. The main usecase I have in mind is video-stabilization, where one would assign a 2D affine to every frame of a movie.

On a similar topic, if I pick one transform object (e.g. one affine matrix transform in the chain), is that intended to be a transform of spatial dimensions only? In the example described, we have a spatial affine matrix per time point, but we can't transform data across across time points. In other cases, we might expect all time points to share the same transform. Similarly with channel dimensions, we might always expect all channels to share the same transform.

I bring this up because there are a few bugs related to this with the current system and the lack of explicit dimension names/types (e.g. #1904, #2487, #2348). One thing I'd to see in a doc that comes out of this is the scope of this work regarding dimension names/types.

**Comment by andy-sweet** [comment]
The grid space is one I'd also like to clarify in a doc. I see the intent of that space/transform as one to allow viewing multiple layers side-by-side for easy comparison. When we have one canvas, we need it to achieve that. But if we have multiple canvases, do we still need it? Another related question is if there's a use for tiling beyond 2D?

On main, I think it's quite easy to break grid mode because the transform required to separate layers in world coordinates is dependent on both the data in those layers and the transforms in the preceding part of the chain.

**Comment by andy-sweet** [comment]
> * provides VisPy glsl map and imap code,

The specificity of glsl here seems to imply that we will support using these transforms for visualization, but maybe not for general purpose processing. For example, if I write a napari plugin to find transforms that align some medical images to a common template space and then want to resample those images into that space, would napari provide a way for me to do that (maybe doing GPGPU behind the scenes)? Or would I have to work with the shader code and implement that myself?

**Comment by andy-sweet** [comment]
Last comment on this particular brain dump, but I also made [a list of issues related to transforms](https://hackmd.io/ciRyBvssT66BJD-Sz_5wsw#Transforms-and-coordinate-spaces). As we know, there's also some overlap with [issues relating to layer data and slicing](https://hackmd.io/ciRyBvssT66BJD-Sz_5wsw#Transforms-and-coordinate-spaces).

**Comment by jojoelfe** [comment]
> In the example described, we have a spatial affine matrix per time point, but we can't transform data across across time points.

There is a transform across the time points, it's just not a linear one, because each data point has it's own transform. In an extreme case, a transform could just by a (n+1)D matrix that contains a displacement for each pixel/voxel/datapoint. It would be kind of neat if napari would support that, because users could supply any arbitrary transform. Of course it would be a real pain (or impossible?) to inverse and to calculate bounds.

**Comment by jni** [comment]
Hi all, just another external reference, a fun blogpost from mapbox about dealing with geographic transforms:

https://www.mapbox.com/blog/adaptive-projections

I suspect some of the techniques will come in handy sooner or later. 😊

**Comment by jni** [comment]
Relevant document: spaces and transformations notes in OME-NGFF:

https://github.com/bogovicj/ngff/wiki/Transforms-notes,-examples,-proposals

**Comment by jni** [comment]
This paper was just published at JOSS: [ndcube: Manipulating N-dimensional Astronomical Data in Python](https://joss.theoj.org/papers/10.21105/joss.05296). It's basically an xarray-lite type thing for storing both ndarrays *and* their corresponding coordinate transformations based on the World Coordinate System that is apparently universal (?) in astronomy.

I still think that we should use xarray+flexible indices as our data structure, *but* there's some interesting ideas in this paper/implementation that we can keep in mind, e.g.:

> `GlobalCoords` supports scalar coordinates that apply to the whole `NDCube` rather than any of its axes, for example, the timestamp of a 2-D image.

It would be good if the whole scientific Python community could settle on a single "annotated array" type, rather than invent one per field, as seems to be happening. 😅 As we have discussed many times before, there's no better candidate for this than xarray at the moment.

**Comment by jni** [comment]
> It would be good if the whole scientific Python community could settle on a single "annotated array" type

gdi: [EuroPython 2023: Scipp: multi-dimensional arrays with labeled dimensions and physical units](https://youtu.be/4ozM506fJ58) 😂 😭 anyway, that's just gone on my watch-list. 😂

**Comment by Czaki** [comment]
As I could check xarray currently does not support by default dast and tensorstore as data backends. So such a change will require contributions to this project. 

Also will limit the ability to use anything that supports array protocol for data storage. 

But there are advantages.
